### ISSUE TITLE
testng-results.xml order not preserved in junitreports/TEST-*.xml

### ISSUE THREAD

**Issue by mauriziokg** [testng-team/testng#421]
Today I migrated one of our projects from junit4 to testNG (version testng-6.8.5). Since I prefer the keep the layout of junitreport I use the junitreport ant task to convert the junitreports/TEST-_.xml file generated by testNG into a standard junit report. I noticed that the order in the resulting junit report was wrong, even though I had observed during execution that the tests had been run in the right order. When investigating upon this, I found that test-results.xml listed the tests in the real execution order (which in my case is also the alphabetical order btw), but the corresponding junitreports/TEST-_.xml file had the wrong order (random as it seems). I checked lots of documentation, problem reports, etc. but could not find this problem so I create a new ticket (sorry if I overlooked something).

**Comment by juherr** [comment]
Investigate and fix TestNG issue #421:
https://github.com/testng-team/testng/issues/421

The issue is about JUnitReportReporter not preserving test execution order in
junitreports/TEST-*.xml.

Important context:

- PR #1270 fixed #1262 by sorting JUnit report results by @Test priority.
- The current JUnitReportReporter still does:
    Comparator.comparingInt(o -> o.getMethod().getPriority())
- The source results ultimately come from ResultMap, whose storage is a
  ConcurrentHashMap.newKeySet(), so iteration order is not execution order.
- Therefore results having the same priority can still appear in an arbitrary
  order.
- XMLSuiteResultWriter, used for testng-results.xml, explicitly orders results
  chronologically by ITestResult.getStartMillis().
- The existing regression test
  testTestCaseOrderingInJUnitReportReporterWhenPrioritiesDefined()
  only covers distinct priorities and does not cover #421.

Mandatory regression-first workflow:

1. Create the smallest possible reproducer for #421 using multiple test methods
   with the same/default priority and a deterministic execution order.
2. Add a TestNG regression test around JUnitReportReporter that asserts that the
   generated TEST-*.xml testcase order matches the actual execution order.
3. Run this regression test against the current code.
4. Confirm that it fails for the exact ordering problem described in #421.
5. Only after obtaining a red regression test, modify production code.
6. Prefer deriving report order from actual execution information
   (e.g. ITestResult timing/order) rather than inventing another unrelated sort.
   Preserve the behavior covered by #1262 for explicit priorities.
7. Re-run the exact regression test and confirm it is green.
8. Run the existing JUnitReportsTest suite, especially the #1262 priority-order
   regression, plus the relevant Gradle checks.

If the regression test does NOT fail on the current implementation:
- do not modify production code;
- investigate why the ConcurrentHashMap-backed result set nevertheless produces
  deterministic execution order in this scenario;
- try a reproducer that deliberately makes declaration order differ from
  execution order;
- document the attempted reproducer and conclude that #421 is not currently
  reproduced only if supported by concrete test evidence.
